JustSaying/AwsTools/MessageHandling/SqsNotificationListener.cs

```diff
@@ -86,22 +85,16 @@ public SqsNotificationListener WithMessageProcessingStrategy(IMessageProcessingS
             _handlerMap.Add(typeof(T), handlerFunc);
         }
 
-        public void Listen()
+        public void Listen(CancellationToken cancellationToken = default)
         {
             var queue = _queue.QueueName;
             var region = _queue.Region.SystemName;
             var queueInfo = $"Queue: {queue}, Region: {region}";
 
-            _cts = new CancellationTokenSource();
-            Task.Factory.StartNew(async () =>
-                {
-                    while (!_cts.IsCancellationRequested)
-                    {
-                        await ListenLoop(_cts.Token).ConfigureAwait(false);
-                    }
-                })
+            Task.Factory.StartNew(async () => { await ListenLoop(cancellationToken).ConfigureAwait(false); },
```

Comment:
I think we should be passing the `TaskCreationOptions.LongRunning` enum here.

Reply:
We should yes, there is an analyser warning that says the same. But I think it's separate to this PR about `CancellationToken` and `IDisposable`.

Reply:
I disagree on this one, it's an async loop with very short lived thread-bound work, perfect for the thread-pool.

Reply:
@slang25 Huh? This task runs forever (or until cancellation is requested). It's the tasks inside it for each message handler that are short-lived. Or have I got this wrong?

Reply:
See [line 249](https://github.com/justeat/JustSaying/pull/403/files#diff-9005575960680ff0f5e43cfda505b5a2R249)

Reply:
The loop should be refactored any to remove the `ContinueWith` - I think that the threading should be reviewed in a separate PR.

Reply:
The `Task` never completes, but it's not a thread bound thing that runs forever. Its a leaky abstraction I guess. See this: https://twitter.com/davidfowl/status/1049198826668486657?s=19

Reply:
I don't think newing up a thread solves anything. What would you do? Have a synchronous loop that blocks each Listen iteration? It's not solving any problems.

If there was part of the loop that was blocking for a non-ideal period of time, then we'd need to think about how to solve it. But as it is, this is a perfect use case for the thread pool.

Think of the loop as a dashed line, where the solid part is the thread bound part. This is 1 line, with massive gaps in (not too important), but the solid parts are really short (very important).

Reply:
I've just concluded that "TaskCreationOptions.LongRunning is useless for Task.Run with async/await", and then found this blog post: 
http://blog.i3arnon.com/2015/07/02/task-run-long-running/

Reply:
@slang25 Thanks. That's cleared it right up for me!

Reply:
Update on this; this has now been clarified in the async guidance (see the 3rd lightbulb):
https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/7e773b7ac2e266c1368113c63db5a8a3b2f5334d/AsyncGuidance.md#avoid-using-taskrun-for-long-running-work-that-blocks-the-thread